include/deal.II/trilinos/nox.templates.h

```diff
@@ -0,0 +1,1027 @@
+
+              if (f.vector)
+                *new_group->f.vector = *f.vector;
+
+              if (gradient.vector)
+                *new_group->gradient.vector = *gradient.vector;
+
+              if (newton.vector)
+                *new_group->newton.vector = *newton.vector;
+
+              new_group->is_valid_f = is_valid_f;
+              new_group->is_valid_j = is_valid_j;
+            }
+          else
+            Assert(copy_type == NOX::CopyType::ShapeCopy, ExcInternalError());
+
+          return new_group;
+        }
+
+        /**
+         * Compute the Newton direction, using parameters for the linear solve.
+         */
+        NOX::Abstract::Group::ReturnType
+        computeNewton(Teuchos::ParameterList &p) override
+        {
+          if (isNewton())
+            return NOX::Abstract::Group::Ok;
+
+          if (isF() == false || isJacobian() == false)
+            return NOX::Abstract::Group::BadDependency;
+
+          if (newton.vector == nullptr)
+            newton.vector = std::make_shared<VectorType>();
+
+          newton.vector->reinit(*f.vector, false);
+
+          const double tolerance = p.get<double>("Tolerance");
+
+          if (solve_with_jacobian(*f.vector, *newton.vector, tolerance) != 0)
+            return NOX::Abstract::Group::NotConverged;
+
+          newton.scale(-1.0);
+
+          return NOX::Abstract::Group::Ok;
+        }
+
+        /**
+         * Applies Jacobian to the given input vector and puts
+         * the answer in the result.
+         */
+        NOX::Abstract::Group::ReturnType
+        applyJacobian(const NOX::Abstract::Vector &input,
+                      NOX::Abstract::Vector &      result) const override
+        {
+          if (apply_jacobian == nullptr)
+            return NOX::Abstract::Group::NotDefined;
+
+          if (!isJacobian())
+            return NOX::Abstract::Group::BadDependency;
+
+          const auto *input_ = dynamic_cast<const Vector<VectorType> *>(&input);
+          const auto *result_ =
+            dynamic_cast<const Vector<VectorType> *>(&result);
+
+          if (apply_jacobian(*input_->vector, *result_->vector) != 0)
+            return NOX::Abstract::Group::Failed;
+
+          return NOX::Abstract::Group::Ok;
+        }
+
+      private:
+        /**
+         * Reset state.
+         */
+        void
+        reset()
+        {
+          is_valid_f = false;
+          is_valid_j = false;
+        }
+
+        // internal vectors
+        Vector<VectorType> x, f, gradient, newton;
+
+        // helper functions to compute residual, to setup jacobian, and
+        // solve jacobian
+        std::function<int(const VectorType &, VectorType &)> residual;
+        std::function<int(const VectorType &)>               setup_jacobian;
+        std::function<int(const VectorType &, VectorType &)> apply_jacobian;
+        std::function<int(const VectorType &, VectorType &, const double)>
+          solve_with_jacobian;
```

Comment:
In the context `residual(*x.vector, *f.vector)` it is suggested that the first argument represents the total solution / linearisation point, rather than the solution update for the timestep or the previous newton increment. Although the NOX documentation probably states this, I think that it would be worth stating this somewhere as nonlinear solution schemes could be formulated in terms of any one of these.